docs/design/rosbag2_snapshot_recorder/rosbag2_snapshot_recorder.md

```diff
@@ -0,0 +1,738 @@
++    {
++      writer_->snapshot();
++      return true;
++    }
++  return false;
++  }
+
+private:
++  /// Add service to handle snapshot functionality
++  rclcpp::Service<rosbag2_interfaces::srv::Snapshot>::SharedPtr snapshot_srv;
+}
+}
+}
+
+
+```
+
+<!-- ### Flushing fix proposal
+
+New idea to improve snapshotting is to create separate `CacheConsumer` and `MessageCache` classes (with different names but sharing same interfaces/deriving most functionality). It all comes to the fact that flushing prevents `MessageCache` from pushing messages to `MessageCacheBuffer` while bag splitting. Moreover every push ends with `notify_buffer_consumer()` which triggers buffer swap, but when snapshotting we want to switch buffers only when saving bagfile to disk. That is why I would opt for changing `CacheConsumer::close()` and `CacheConsumer::~CacheConsumer` action to more suitable for snapshotting.
+
+```diff
+
+void MessageCache::push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg)
+{
+  (...)
+-  notify_buffer_consumer();
+}
+```
+
+This change would distinguish calling `SequentialWriter::switch_to_next_storage()` (as snapshot) and `SequentialWriter::reset()` (which is called only in destructor).Additionally we would acquire data the whole time during closing one and opening another storage. It would require some more work to apply classes deriviation, unify interfaces and develop mechanism for objects creation.
+
+Presented idea won't work in normal bag recording (and snapshotting) due to writing closed storage object inside consumer thread during bag split. In snapshot mode even program close (Ctrl-C) won't break the program behavior due to `MessageCache::flushing_` flag set when closing program and proper mutex locking during pushing and waiting for buffer swap. -->
+
+### New Services
+
+rosbag2_interfaces/srv/Snapshot
+
+```
+---
+```
+
+rosbag2_interfaces/srv/SetMemory
+
+```
+float64 memory
+---
+bool success  # true if memory value is positive or non-positive but duration is positive
+```
+
+rosbag2_interfaces/srv/SetDuration
+
+```
+float64 duration
+---
+bool success  # true if memory value is positive or non-positive but memory is positive
+```
+
+rosbag2_interfaces/srv/ClearBuffer
+
+```
+---
+```
+
+rosbag2_interfaces/srv/Pause and rosbag2_interfaces/srv/Resume are shared with `Player` node.
+
+### Feature implementation tests
+
+* `CacheBufferFactory` check if returns good `CacheBufferInterface` type instances, dependent on passed arguments
+* `MessageSnapshotBuffer` check pushing behavior
+* `MessageCache` check synchronization and start timestamp calculation
+* `SequentialWriter` test if snapshot forces using cache, starting timestamp change in snapshot mode and bag split when calling snapshot
+* `Recorder` service call check
+
+#### TODO
+
+* Deleting empty rosbag after node closure
+* Resolve how to implement pause/resume functionality
+* Think about configuration change during runtime
```

Comment:
I think that this feature can be intentionally excluded from this design. Maybe add an "Out of scope" section near the top, something like


```
## Out of scope

* Reconfiguration of recording settings during runtime
```